main/output.c

```diff
@@ -962,16 +969,37 @@ static inline php_output_handler_status_t php_output_handler_op(php_output_handl
 			handler->func.user->fci.retval = &retval;
 
 			if (SUCCESS == zend_call_function(&handler->func.user->fci, &handler->func.user->fcc) && Z_TYPE(retval) != IS_UNDEF) {
-				if (Z_TYPE(retval) != IS_STRING) {
+				if (Z_TYPE(retval) != IS_STRING || handler->flags & PHP_OUTPUT_HANDLER_PRODUCED_OUTPUT) {
 					// Make sure that we don't get lost in the current output buffer
 					// by disabling it
 					handler->flags |= PHP_OUTPUT_HANDLER_DISABLED;
-					php_error_docref(
-						NULL,
-						E_DEPRECATED,
-						"Returning a non-string result from user output handler %s is deprecated",
-						ZSTR_VAL(handler->name)
-					);
+					// Make sure we keep a reference to the handler name in
+					// case
+					// * The handler produced output *and* returned a non-string
+					// * The first deprecation message causes the handler to
+					// be removed
+					zend_string *handler_name = handler->name;
+					zend_string_addref(handler_name);
```

Comment:
This would be a little cleaner as `zend_string *handler_name = zend_string_copy(handler->name);`